inttestutils/transfer.go

```diff
@@ -0,0 +1,240 @@
+// t - The testing object
+func CountArtifactsInPath(pattern string, serverDetails *config.ServerDetails, t *testing.T) int {
+	searchCmd := generic.NewSearchCommand()
+	searchCmd.SetServerDetails(serverDetails).SetSpec(spec.NewBuilder().Pattern(pattern).BuildSpec())
+	reader, err := searchCmd.Search()
+	assert.NoError(t, err)
+	defer assert.NoError(t, reader.Close())
+	lenght, err := reader.Length()
+	assert.NoError(t, err)
+	return lenght
+}
+
+// Wait for a metadata file to automatically generated in Artifactory
+// pattern - The search pattern
+// serverDetails - The Artifactory server details
+// t - The testing object
+func WaitForCreationInArtifactory(pattern string, serverDetails *config.ServerDetails, t *testing.T) {
+	searchCmd := generic.NewSearchCommand()
+	searchCmd.SetServerDetails(serverDetails).SetSpec(spec.NewBuilder().Pattern(pattern).BuildSpec())
+	for i := 0; i < 20; i++ {
+		reader, err := searchCmd.Search()
+		assert.NoError(t, err)
+		defer assert.NoError(t, reader.Close())
+		if !reader.IsEmpty() {
+			return
+		}
+		time.Sleep(5 * time.Second)
+	}
+	assert.Fail(t, "Couldn't find in target Artifactory: "+pattern)
+}
+
+// Asynchronously execute transfer-files
+// artifactoryCli - Source Artifactory CLI
+// wg - Wait group to update when done
+// filesTransferFinished - Changes to true when the file transfer process finished
+// t - The testing object
+func AsyncExecTransferFiles(artifactoryCli *tests.JfrogCli, wg *sync.WaitGroup, filesTransferFinished *bool, t *testing.T) {
+	wg.Add(1)
+	go func() {
+		defer func() {
+			wg.Done()
+			*filesTransferFinished = true
+		}()
+		// Execute transfer-files
+		assert.NoError(t, artifactoryCli.WithoutCredentials().Exec("transfer-files", "default", "target", "--include-repos="+tests.RtRepo1))
+	}()
+}
+
+// Asynchronously update the number of threads in transfer-files command
+// wg - Wait group to update when done
+// filesTransferFinished - Changes to true when the file transfer process finished
+// t - The testing object
+func AsyncUpdateThreads(wg *sync.WaitGroup, filesTransferFinished *bool, t *testing.T) {
+	wg.Add(1)
+	go func() {
+		threadsCount := 0
+		defer wg.Done()
+
+		// Wait for the number of threads to be updated to the non-zero default and increace the number of threads by 1
+		for !*filesTransferFinished {
+			threadsCount = transferfiles.GetThreads()
+			if threadsCount == 0 {
+				time.Sleep(time.Second)
+				continue
+			}
+			conf := &utils.TransferSettings{ThreadsNumber: threadsCount + 1}
+			assert.NoError(t, utils.SaveTransferSettings(conf))
+			break
+		}
+
+		// Wait for the number of threads to be increase by 1
+		threadCountChanged := false
+		for !threadCountChanged && !*filesTransferFinished {
+			if transferfiles.GetThreads() == threadsCount+1 {
+				threadCountChanged = true
+				break
+			}
+			time.Sleep(time.Second)
+		}
+
+		// If false, the transfer-files process is finished before the threads changed
```

Comment:
This could be a race. We don't have to fix it now, but we might run into it once the 15 idle seconds are reduced.

Reply:
It is a race, but a race between a rabbit and a turtle. This code runs while uploading 101 files. If needed in the future, we can change 101 to a bigger number, although I'm skeptical.